### PR TITLE
feat!: define wrappers around package that point to internals

### LINKED ISSUE #1561: py: Add a "Pointed package" protocol
Guppy can compile functions into hugr packages, which have a function identified as an entry point.
This is similar to tket2 circuits, which point to a region in a hugr.

We should unify this notion in a Protocol in `hugr-py`, so both libraries can interact without depending on each other.
```python
class Proto(Protocol):
    def package(self) -> Package: ...
    def hugr_index(self) -> int: ...
    def root(self) -> Node: ...
```

### PR DESCRIPTION
**Pull request by ss2165**
Closes #1561

So far just top level things, can add inner ones (cfg, blocks, etc.) later.

BREAKING CHANGE: `Package` moved to new `hugr.package` module

### CODE REVIEW

File: hugr-py/src/hugr/package.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    def module(self) -> Hugr:
+        """Hugr definition of the module."""
+        return self.package.modules[self.module_index]
+
+    def to_executable_package(self) -> "ExecutablePackage":
+        """Create an executable package from a module containing a main function.
+
+        Raises:
+            StopIteration: If the module does not contain a main function.
```
Comment (aborgna-q) [inline hugr-py/src/hugr/package.py:71]: Rethrow a more appropriate exception, with a custom message.
Either a custom one, or `ValueError`

File: hugr-py/src/hugr/package.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+class ModulePointer(PackagePointer):
+    """Pointer to a module in a package."""
+
+    module_index: int
+
+    def module(self) -> Hugr:
+        """Hugr definition of the module."""
+        return self.package.modules[self.module_index]
+
+    def to_executable_package(self) -> "ExecutablePackage":
```
Comment (aborgna-q) [inline hugr-py/src/hugr/package.py:67]:
```python
from __future__ import annotations
```
and remove the quotes

File: hugr-py/src/hugr/package.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+@dataclass
+class PackagePointer:
+    """Classes that point to packages and their inner contents."""
+
+    package: Package
+
+    def get_package(self) -> Package:
+        """Get the package pointed to."""
+        return self.package
```
Comment (mark-koch) [inline hugr-py/src/hugr/package.py:54]: What is the purpose of this method?

Comment (ss2165) [reply hugr-py/src/hugr/package.py:54]: good point!

Comment (ss2165) [reply hugr-py/src/hugr/package.py:54]: (vestigial from original issue which proposed protocols)

File: hugr-py/src/hugr/package.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+@dataclass
+class ModulePointer(PackagePointer):
+    """Pointer to a module in a package."""
+
+    module_index: int
+
+    def module(self) -> Hugr:
+        """Hugr definition of the module."""
+        return self.package.modules[self.module_index]
```
Comment (mark-koch) [inline hugr-py/src/hugr/package.py:65]: This could be a property (same for all other methods in this PR)

File: hugr-py/src/hugr/package.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+class FuncDefnPointer(NodePointer[FuncDefn]):
+    """Pointer to a function definition in a module."""
+
+    def func_defn(self) -> FuncDefn:
+        """Function definition."""
+        return self.node_op()
+
+
+@dataclass
+class ExecutablePackage(FuncDefnPointer):
```
Comment (aborgna-q) [inline hugr-py/src/hugr/package.py:127]:
```suggestion
class ExecutablePackage(FuncDefnPointer):
        """PackagePointer with a defined entrypoint node."""
        
```

File: hugr-py/src/hugr/package.py
Comment (aborgna-q) [inline hugr-py/src/hugr/package.py:1]: It'd be useful to add 
```py
    Args:
        package: ...
        module: ...
        ...
```
sections to the dataclass docs.
Python doesn't make too clear what attributes are getting inherited in the constructor.

**Review by aborgna-q — APPROVED** [review]
Nice.

The followup question will now be how do we serialize these :)

**Pull request merged**